File: src/main/java/abstract_factory/product/Vehicle.java
package abstract_factory.product;

import abstract_factory.product.part.Engine;
import abstract_factory.product.part.Option;
import abstract_factory.product.part.Tire;

import java.util.List;

public abstract class Vehicle {
    protected String name;
    protected Engine engine;
    protected Tire tire;
    protected List<Option> options;

    public abstract void assembly();

    public void loadShip() {
        System.out.println("배에 선적");
    }

    public Engine getEngine() {
        return engine;
    }

    public Tire getTire() {
        return tire;
    }

    public List<Option> getOptions() {
        return options;
    }

    public String getName() {
        return name;
    }
}

File: kotlin/src/main/kotlin/abstract_factory/product/Vehicle.kt
package abstract_factory.product

import abstract_factory.product.part.Engine
import abstract_factory.product.part.Option
import abstract_factory.product.part.Tire

abstract class Vehicle {

    abstract var name: String
    lateinit var engine: Engine
    lateinit var tire: Tire
    lateinit var options: List<Option?>

    abstract fun assembly()

    fun loadShip() {
        println("배에 선적")
    }
}